File: internal/app/manual.go
package app

import (
	"math"
	"strings"

	"github.com/ProfessorBagholder/Bagholder/internal/py"
	"github.com/ProfessorBagholder/Bagholder/internal/store"
	"github.com/ProfessorBagholder/Bagholder/internal/ws"
)

func upper(v any) string { return strings.ToUpper(py.Strip(py.S(v))) }

func pick(body map[string]any, keys ...string) any {
	for _, k := range keys {
		if v, ok := body[k]; ok && v != nil {
			return v
		}
	}
	return nil
}

func firstS(body map[string]any, keys ...string) string {
	for _, k := range keys {
		if v := py.S(body[k]); v != "" {
			return v
		}
	}
	return ""
}

func manualFromFields(body map[string]any) store.Activity {
	side := upper(pick(body, "side"))
	if side == "" {
		side = "BUY"
	}
	if side != "BUY" && side != "SELL" {
		side = "BUY"
	}
	qty := math.Abs(py.Num(pick(body, "qty", "quantity"), 0))
	px := math.Abs(py.Num(pick(body, "price", "unitPrice"), 0))
	date := ws.DateOnly(firstS(body, "date", "transactionDate", "occurredAt"))
	if date == "" {
		date = today()
	}
	symbol := upper(body["symbol"])
	currency := upper(body["currency"])
	if currency == "" {
		currency = "CAD"
	}
	if currency != "CAD" && currency != "USD" {
		currency = "CAD"
	}
	accountID := firstS(body, "accountId", "account")
	if accountID == "" {
		accountID = "manual"
	}
	signedQty := qty
	cash := -(qty * px)
	direction := "DEBIT"
	desc := "Buy"
	if side == "SELL" {
		signedQty = -qty
		cash = qty * px
		direction = "CREDIT"
		desc = "Sell"
	}
	if symbol != "" {
		desc += " " + py.G(qty) + " " + symbol + " @ " + py.G(px)
	}
	accountType := py.S(body["accountType"])
	if accountType == "" && accountID == "manual" {
		accountType = "Manual"
	}
	return store.Activity{ID: py.UUID4(), OccurredAt: date, TransactionDate: date, SettlementDate: date, AccountID: accountID, BookID: accountID, AccountType: accountType,
		ActivityType: "Trade", ActivitySubType: side, Description: desc, Direction: direction, Symbol: symbol, Name: symbol, Currency: currency, Quantity: signedQty,
		UnitPrice: px, Commission: math.Abs(py.Num(body["commission"], 0)), NetCashAmount: cash, Category: "trade", Source: "manual"}
}

func activityFromMap(m map[string]any) store.Activity {
	a := store.Activity{ID: py.S(m["id"]), CanonicalID: firstS(m, "canonicalId", "canonical_id"), OccurredAt: py.S(m["occurredAt"]), TransactionDate: py.S(m["transactionDate"]),
		SettlementDate: py.S(m["settlementDate"]), AccountID: py.S(m["accountId"]), BookID: py.S(m["bookId"]), FifoID: py.S(m["fifoId"]), AccountType: py.S(m["accountType"]),
		ActivityType: py.S(m["activityType"]), ActivitySubType: py.S(m["activitySubType"]), Description: py.S(m["description"]), Direction: py.S(m["direction"]),
		Symbol: py.S(m["symbol"]), Name: py.S(m["name"]), Currency: py.S(m["currency"]), Quantity: py.Num(m["quantity"], 0), UnitPrice: py.Num(m["unitPrice"], 0),
		Commission: py.Num(m["commission"], 0), NetCashAmount: py.Num(m["netCashAmount"], 0), Category: py.S(m["category"]), Source: py.S(m["source"]),
		RawType: py.S(m["rawType"]), AftType: py.S(m["aftType"]), CounterSymbol: py.S(m["counterSymbol"]), SecurityID: py.S(m["securityId"])}
	if v, ok := py.NumOK(m["balance"]); ok {
		a.Balance = &v
	}
	return a
}

func normalizeLocalRow(act store.Activity) store.Activity {
	source := act.Source
	if source == "" {
		source = "manual"
	}
	if source == "wealthsimple" {
		if cid := store.CanonicalFromRow(&act, "wealthsimple"); cid != "" {
			act.CanonicalID = cid
			act.Source = "wealthsimple"
			return act
		}
		source = "manual"
	}
	act.Source = source
	act.CanonicalID = ""
	if act.AccountID == "" {
		act.AccountID = "manual"
	}
	if act.BookID == "" {
		act.BookID = act.AccountID
	}
	if act.ID == "" || store.LooksLikeHomemadeID(act.ID) {
		act.ID = py.UUID4()
	}
	if act.OccurredAt == "" {
		act.OccurredAt = act.TransactionDate
	}
	return act
}

func (a *App) appendManual(body map[string]any) map[string]any {
	if body == nil {
		body = map[string]any{}
	}
	var rows []store.Activity
	if list, ok := body["activities"].([]any); ok {
		for _, r := range list {
			if m, ok := r.(map[string]any); ok {
				rows = append(rows, activityFromMap(m))
			}
		}
	} else if m, ok := body["activity"].(map[string]any); ok && len(m) > 0 {
		rows = append(rows, activityFromMap(m))
	} else {
		rows = append(rows, manualFromFields(body))
	}
	for i := range rows {
		rows[i] = normalizeLocalRow(rows[i])
	}
	result := a.st.MergeLocalRows(rows)
	snap := a.st.Snapshot(false)
	if snap.SyncedAt == "" {
		a.st.SetMeta("synced_at", nowStamp())
		snap = a.st.Snapshot(false)
	}
	a.mu.Lock()
	if snap.SyncedAt != "" {
		a.state.lastSync = snap.SyncedAt
	}
	a.mu.Unlock()
	saved := result.Activities
	out := map[string]any{"ok": true, "added": result.Added, "duplicates": result.Duplicates}
	switch {
	case len(saved) == 1:
		out["activity"] = saved[0]
	case len(saved) > 0:
		out["activities"] = saved
	case len(rows) == 1:
		out["activity"] = rows[0]
	}
	return out
}

File: rust/crates/server/src/orders/manual.rs
//! A fill the person enters by hand.

use super::*;
use bagholder_store::activities::ActivityRow;
use serde::Serialize;

// ---------------------------------------------------------------------------
// manual activity
// ---------------------------------------------------------------------------

/// `POST /api/book/append`'s body: either whole rows (`activities`, or the
/// singular `activity`), or the ticket-shaped fields a hand-entered trade
/// sends -- each with the alternative spellings the page and the phones have
/// used, read in the precedence `manual_from_fields` resolves.
#[derive(Clone, Debug, Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct BookAppend {
    #[serde(deserialize_with = "bagholder_model::lenient::list")]
    pub activities: Vec<ActivityRow>,
    pub activity: Option<ActivityRow>,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub side: String,
    #[serde(deserialize_with = "bagholder_model::lenient::maybe_number")]
    pub qty: Option<f64>,
    #[serde(deserialize_with = "bagholder_model::lenient::maybe_number")]
    pub quantity: Option<f64>,
    #[serde(deserialize_with = "bagholder_model::lenient::maybe_number")]
    pub price: Option<f64>,
    #[serde(deserialize_with = "bagholder_model::lenient::maybe_number")]
    pub unit_price: Option<f64>,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub date: String,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub transaction_date: String,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub occurred_at: String,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub symbol: String,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub currency: String,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub account_id: String,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub account: String,
    #[serde(deserialize_with = "bagholder_model::lenient::text")]
    pub account_type: String,
    #[serde(deserialize_with = "bagholder_model::lenient::maybe_number")]
    pub commission: Option<f64>,
}

fn first_nonempty(vs: &[&str]) -> String {
    vs.iter().find(|v| !v.is_empty()).map(|v| v.to_string()).unwrap_or_default()
}

fn date_only_str(t: &str) -> String {
    let t = t.trim();
    if t.is_empty() {
        return String::new();
    }
    t.split('T').next().unwrap_or("").chars().take(10).collect()
}

pub(super) fn manual_from_fields(body: &BookAppend) -> ActivityRow {
    let mut side = body.side.trim().to_uppercase();
    if side != "BUY" && side != "SELL" {
        side = "BUY".into();
    }
    let qty = body.qty.or(body.quantity).unwrap_or(0.0).abs();
    let px = body.price.or(body.unit_price).unwrap_or(0.0).abs();
    let mut date = date_only_str(&first_nonempty(&[&body.date, &body.transaction_date, &body.occurred_at]));
    if date.is_empty() {
        date = crate::app::today_utc();
    }
    let symbol = body.symbol.trim().to_uppercase();
    let mut currency = body.currency.trim().to_uppercase();
    if currency.is_empty() {
        currency = "CAD".into();
    }
    if currency != "CAD" && currency != "USD" {
        currency = "CAD".into();
    }
    let mut account_id = first_nonempty(&[&body.account_id, &body.account]);
    if account_id.is_empty() {
        account_id = "manual".into();
    }
    let buy = side == "BUY";
    let signed_qty = if buy { qty } else { -qty };
    let cash = if buy { -(qty * px) } else { qty * px };
    let account_type = if !body.account_type.is_empty() {
        body.account_type.clone()
    } else if account_id == "manual" {
        "Manual".into()
    } else {
        String::new()
    };
    let desc = format!(
        "{}{}",
        if buy { "Buy" } else { "Sell" },
        if symbol.is_empty() { String::new() } else { format!(" {} {} @ {}", fmt_g(qty), symbol, fmt_g(px)) }
    );
    ActivityRow {
        id: uuid4(),
        occurred_at: date.clone(),
        transaction_date: date.clone(),
        settlement_date: date,
        account_id: account_id.clone(),
        book_id: account_id,
        account_type,
        activity_type: "Trade".into(),
        activity_sub_type: side,
        description: desc,
        direction: if buy { "DEBIT" } else { "CREDIT" }.into(),
        symbol: symbol.clone(),
        name: symbol,
        currency,
        quantity: signed_qty,
        unit_price: px,
        commission: body.commission.unwrap_or(0.0).abs(),
        net_cash_amount: cash,
        category: "trade".into(),
        source: "manual".into(),
        ..ActivityRow::default()
    }
}

pub(super) fn normalize_local_row(act: ActivityRow) -> ActivityRow {
    let mut act = act;
    let mut source = act.source.clone();
    if source.is_empty() {
        source = "manual".into();
    }
    if source == "wealthsimple" {
        if let Some(cid) = bagholder_store::activities::canonical_from_row(&act, "wealthsimple") {
            act.canonical_id = Some(cid);
            act.source = "wealthsimple".into();
            return act;
        }
        source = "manual".into();
    }
    act.source = source;
    act.canonical_id = None;
    if act.account_id.is_empty() {
        act.account_id = "manual".into();
    }
    if act.book_id.is_empty() {
        act.book_id = act.account_id.clone();
    }
    if act.id.is_empty() || bagholder_store::activities::looks_like_homemade_id(&act.id) {
        act.id = uuid4();
    }
    if act.occurred_at.is_empty() {
        act.occurred_at = act.transaction_date.clone();
    }
    act
}

/// `POST /api/book/append`'s answer: what was added and, when it is one row
/// or the row the person's own ticket described, that row back.
#[derive(Debug, Clone, Serialize, ts_rs::TS)]
#[serde(rename_all = "camelCase")]
pub struct Appended {
    pub ok: bool,
    pub added: usize,
    pub duplicates: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub activity: Option<ActivityRow>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub activities: Option<Vec<ActivityRow>>,
}

pub fn append_manual(app: &Arc<App>, body: &BookAppend) -> Value {
    let rows: Vec<ActivityRow> = if !body.activities.is_empty() {
        body.activities.clone()
    } else if let Some(a) = &body.activity {
        vec![a.clone()]
    } else {
        vec![manual_from_fields(body)]
    };
    let rows: Vec<ActivityRow> = rows.into_iter().map(normalize_local_row).collect();
    let conn = db(app);
    let result = must(bagholder_store::merge::merge_local_rows(&conn, &rows, &uuid4));
    let mut snap = snapshot(app);
    if !tr(&snap, "syncedAt") {
        let stamp = now_iso();
        must(bagholder_store::tables::set_meta(&conn, "synced_at", &stamp));
        snap = snapshot(app);
    }
    {
        let mut st = app.state.lock().unwrap();
        let synced = f(&snap, "syncedAt");
        if !synced.is_empty() {
            st.last_sync = synced;
        }
    }
    let saved = result.activities;
    let mut out = Appended { ok: true, added: result.added, duplicates: result.duplicates, activity: None, activities: None };
    if saved.len() == 1 {
        out.activity = Some(saved[0].clone());
    } else if !saved.is_empty() {
        out.activities = Some(saved);
    } else if rows.len() == 1 {
        out.activity = Some(rows[0].clone());
    }
    serde_json::to_value(&out).unwrap_or(Value::Null)
}
